### PR TITLE
agent: Refactor command line parsing to use a framework

### PR DESCRIPTION
**Pull request by stevenhorsman**
- Use clap to do command line parsing

Fixes: #3284

Co-authored-by: Samuel Ortiz <samuel.e.ortiz@protonmail.com>
Co-authored-by: stevenhorsman <steven@uk.ibm.com>
    
Signed-off-by: stevenhorsman <steven@uk.ibm.com>

### CODE REVIEW

**Comment by stevenhorsman** [comment]
With these changes the command line interaction look like this:
- A new generated help when `-h/--help` is available:
```bash
target/x86_64-unknown-linux-musl/release/kata-agent -h
```
shows
```bash
kata-agent

USAGE:
    kata-agent [OPTIONS] [SUBCOMMAND]

OPTIONS:
    -c, --config <CONFIG>    Specify a custom agent config file
    -h, --help               Print help information
    -v, --version            Print the version information

SUBCOMMANDS:
    help    Print this message or the help of the given subcommand(s)
    init
```
- Version string unchanged (except `-v` short string is now an option):
```bash
target/x86_64-unknown-linux-musl/release/kata-agent --version
```
shows
```bash
kata-agent version 2.4.0-alpha0 (api version: 0.0.1, commit version: 2.4.0-alpha0-ff572fa2192c051246c5bccb29c968ac597b5360-dirty, type: rust)
```
- Testing the `-c/--config` custom agent config file:
  - Create a sample agent-config file:
  ```
  cat << EOT | tee agent_config.toml
  # This is an agent configuration file example.
  dev_mode = true
  server_addr = 'unix://@/tmp/foo.socket'
  
  [endpoints]
  allowed = [ "CreateContainer", "RemoveContainer", "AddSwap"]
  EOT
  ```
  - Start the agent with the config option (and kill it once the announce is printed):
  ```bash
  target/x86_64-unknown-linux-musl/release/kata-agent -c ./agent_config.toml
  ```
  results in:
  ```bash
  {"msg":"announce","level":"INFO","ts":"2021-12-10T02:25:11.808117234-08:00","name":"kata-agent","pid":"812155","version":"0.1.0","source":"agent","subsystem":"root","agent-version":"2.4.0-alpha0","api-version":"0.0.1","agent-commit":"2.4.0-alpha0-ff572fa2192c051246c5bccb29c968ac597b5360-dirty","config":"AgentConfig { debug_console: false, dev_mode: true, log_level: Info, hotplug_timeout: 3s, debug_console_vport: 0, log_vport: 0, container_pipe_size: 0, server_addr: \"unix://@/tmp/foo.socket\", unified_cgroup_hierarchy: false, tracing: false, endpoints: AgentEndpoints { allowed: {\"RemoveContainer\", \"AddSwap\", \"CreateContainer\"}, all_allowed: false }, supports_seccomp: true }","agent-type":"rust"}
  ```
  - Note: `AgentEndpoints { allowed: {\"RemoveContainer\", \"AddSwap\", \"CreateContainer\"}`, `server_addr: \"unix://@/tmp/foo.socket\"` and `dev_mode: true` are using the config file versions rather than the defaults.

**Comment by stevenhorsman** [comment]
I'm having looks of issues with cargo's interaction with clap, so I'm going to start stripping back this PR to try and isolate if there is any specific code problems, or if clap and cargo test -- --nocapture are just fundamentally incompatible.
What I'd really like to do is leave `config.rs` untouched, but to change `AGENT_CONFIG` in `main.rs` so that if the config file is passed in it is initialised by `AgentConfig::from_config_file(&config_file)` rather than `AgentConfig::from_cmdline("/proc/cmdline")`, but as it is a static ref I'm not sure how I can do that without ending up with unsafe code 😞

**Comment by stevenhorsman** [comment]
The agent build is failing to compile for s390x and ARM, so I guess the `clap` crate's dependencies aren't supported on those platforms.

**Comment by stevenhorsman** [comment]
I've created a native only implementation of this PR in #3261 as the `clap` version seems to cause dependency issues. If that is merged then I'm happy to try the refactor use `clap` once v3 has probably released and see if the situation is better then.

File: src/agent/src/main.rs [hunk 1]
```diff
@@ -94,16 +94,16 @@ struct AgentOpts {
...
     subcmd: Option<SubCommand>,
+    /// Specify a custom agent config file
+    #[clap(short, long)]
+    config: Option<String>,
 }
 
 #[derive(Parser)]
 enum SubCommand {
-    Init(Init),
+    Init {},
```
Comment (dgibson) [inline src/agent/src/main.rs:104]: This change looks like a cleanup which should be folded into the first patch.

**Review by dgibson — COMMENTED** [review]
The clap conversion looks fine, although I do wonder how many extra dependencies this pulls into the agent.  The current draft looks rather odd, of course, since it adds the config option to the clap structure, but then doesn't actually use it from there.

**Comment by stevenhorsman** [comment]
> The current draft looks rather odd, of course, since it adds the config option to the clap structure, but then doesn't actually use it from there.

Yes, I agree. Due the the issues with parsing arguments outside of main (mentioned in https://github.com/kata-containers/kata-containers/pull/3253#issuecomment-991189526) we can use the clap logic for init and version, but not in `config.rs` at the moment, which makes it a bit of a compromise between full clap and native

**Comment by stevenhorsman** [comment]
With the native implementation having merged, this PR is just about the refactor to use the `clap` crate in the `main.rs` parsing. If it's too confusing then I'm happy to re-create this PR to reduce some of the noise.

File: src/agent/Cargo.toml [hunk 2]
```diff
@@ -60,6 +60,7 @@ vsock-exporter = { path = "vsock-exporter" }
 # Configuration
 serde = { version = "1.0.129", features = ["derive"] }
 toml = "0.5.8"
+clap = { version = "3.0.0-rc.4", features = ["derive"] }
```
Comment (jodh-intel) [inline src/agent/Cargo.toml:63]: You could bump this to `3.0.0-rc.7` now.

**Review by jodh-intel — APPROVED** [review]
Thanks @stevenhorsman.

lgtm

File: src/agent/Cargo.toml (hunk 2 above)
Comment (stevenhorsman) [reply src/agent/Cargo.toml]: Thanks, I've done that, so will re-run the tests :)

Comment (stevenhorsman) [reply src/agent/Cargo.toml]: clap `3.0.1` has released now, so I've updated to that and will run the tests again.

**Review by liubin — APPROVED** [review]

**Pull request merged**